### ISSUE TITLE
API disconnect - reboot will not resolve.

### ISSUE THREAD

**Issue by alsoszaa** [ollo69/ha-smartthinq-sensors#401]
**Describe the bug**
If possible attach the JSON info file for your devices.

**Expected behavior**
If applicable, a clear and concise description of what you expected to happen.

**Screenshots**

**Environment details:**
 - Environment (HASSIO, Raspbian, etc):
 - Home Assistant version installed:

Home Assistant 2022.11.2
Supervisor 2022.10.2
Frontend 20221108.0 - latest

 - Component version installed: latest 0.24.2
 - Last know working version:
 - LG device type and model with issue:
 - LG devices connected (list):

**Output of HA logs**
Paste the relavant output of the HA log here.

```

Logger: custom_components.smartthinq_sensors
Source: custom_components/smartthinq_sensors/wideq/core_async.py:327 
Integration: SmartThinQ LGE Sensors (documentation, issues) 
First occurred: 11:12:17 PM (1 occurrences) 
Last logged: 11:12:17 PM

Connection not available. ThinQ platform not ready
Traceback (most recent call last):
  File "/config/custom_components/smartthinq_sensors/__init__.py", line 221, in async_setup_entry
    client = await lge_auth.create_client_from_token(refresh_token, oauth_url)
  File "/config/custom_components/smartthinq_sensors/__init__.py", line 127, in create_client_from_token
    return await ClientAsync.from_token(
  File "/config/custom_components/smartthinq_sensors/wideq/core_async.py", line 1255, in from_token
    await client.refresh()
  File "/config/custom_components/smartthinq_sensors/wideq/core_async.py", line 1196, in refresh
    await self._load_devices()
  File "/config/custom_components/smartthinq_sensors/wideq/core_async.py", line 1112, in _load_devices
    self._devices = await self._session.get_devices()
  File "/config/custom_components/smartthinq_sensors/wideq/core_async.py", line 898, in get_devices
    dashboard = await self.get2("service/application/dashboard")
  File "/config/custom_components/smartthinq_sensors/wideq/core_async.py", line 887, in get2
    return await self._auth.gateway.core.thinq2_get(
  File "/config/custom_components/smartthinq_sensors/wideq/core_async.py", line 283, in thinq2_get
    res = self._manage_lge_result(out, True)
  File "/config/custom_components/smartthinq_sensors/wideq/core_async.py", line 327, in _manage_lge_result
    raise API2_ERRORS[code]()
custom_components.smartthinq_sensors.wideq.core_exceptions.NotConnectedError

```

**Additional context**

**Comment by AndersNielsen85** [comment]
Seems to be some general connection issues for me as well. Wonder if LG changed something in their API?

```
This error originated from a custom integration.

Logger: custom_components.smartthinq_sensors.wideq.device
Source: custom_components/smartthinq_sensors/wideq/core_async.py:329
Integration: SmartThinQ LGE Sensors (documentation, issues)
First occurred: November 30, 2022 at 02:57:43 (3 occurrences)
Last logged: November 30, 2022 at 02:58:03

Unexpected error while updating device status - Device: Vaskemaskine
Unexpected error while updating device status - Device: Tørretumbler
Unexpected error while updating device status - Device: Opvaskemaskine
Traceback (most recent call last):
  File "/config/custom_components/smartthinq_sensors/wideq/device.py", line 155, in refresh
    state = await self.poll(query_device)
  File "/config/custom_components/smartthinq_sensors/wideq/device.py", line 268, in poll
    return await self._poll_v2(query_device)
  File "/config/custom_components/smartthinq_sensors/wideq/device.py", line 295, in _poll_v2
    await self._client.refresh_devices()
  File "/config/custom_components/smartthinq_sensors/wideq/core_async.py", line 1174, in refresh_devices
    await self._load_devices(True)
  File "/config/custom_components/smartthinq_sensors/wideq/core_async.py", line 1112, in _load_devices
    self._devices = await self._session.get_devices()
  File "/config/custom_components/smartthinq_sensors/wideq/core_async.py", line 898, in get_devices
    dashboard = await self.get2("service/application/dashboard")
  File "/config/custom_components/smartthinq_sensors/wideq/core_async.py", line 887, in get2
    return await self._auth.gateway.core.thinq2_get(
  File "/config/custom_components/smartthinq_sensors/wideq/core_async.py", line 283, in thinq2_get
    res = self._manage_lge_result(out, True)
  File "/config/custom_components/smartthinq_sensors/wideq/core_async.py", line 329, in _manage_lge_result
    raise exc.APIError(code, message)
custom_components.smartthinq_sensors.wideq.core_exceptions.APIError

**Comment by ollo69** [comment]
Is this issue still present?

**Comment by djcrawleravp** [comment]
Still happening to me, even the integration works

**Comment by ollo69** [comment]
Are u using last version with last ha release? If yes my suggestion is to try to remove and reconfigure integration.

**Comment by ollo69** [comment]
The problem is that there are 2 way to get `oauth_url` needed for token refresh:

1) use the `oauth url` returned during initial login
2) retrieve `oauth url` dynamically from LG servers

Because 2 methods return 2 different url, I'm not really sure which method is more correct, but for sure in case of first method (that I used until now) if `oauth url` change I will continue using wrong one and this could cause the issue.

I just release a new version that use method 2) hoping that this do not cause issue to any existing installation, please test it and let me know if this solve the issue.

**Comment by djcrawleravp** [comment]
Just checked the new version, the issue is the same, nevertheless it seems only happens once at HA startup, after a couple of seconds the integrations loads and works propperly

**Comment by ollo69** [comment]
> Just checked the new version, the issue is the same, nevertheless it seems only happens once at HA startup, after a couple of seconds the integrations loads and works propperly

This just seems to be an issue related to network connection. Are you using Wi-Fi? Wi-Fi take some times to come up and could be not available immediately during startup.

**Comment by djcrawleravp** [comment]
I’m using latest HA docker version on a RBPi4 on ethernet, everything works pretty well, just this error at startup

**Comment by ollo69** [comment]
Understood, no idea of what could be the reason. Please let me know if you found some more details for this.

**Comment by ollo69** [comment]
@djcrawleravp,

I think that the problem could be related to multiple http requests that runs during HA startup that cause some of them to timeout. I received some times same errors, probably this also depend on how many active integrations and related device you have. In last release I just increment to 15 seconds the default http timeout, please le me know if you still have the error.

**Comment by djcrawleravp** [comment]
@ollo69 Hi, I’ve just tested the new version by the time out seems to persist

![BDC3340E-7A84-40DA-8046-40586FD4B42F](https://user-images.githubusercontent.com/12896341/214434853-1c9adf1e-ffa9-453c-9b81-1e78987d6f41.png)

**Comment by ollo69** [comment]
Honestly I don't want to go over the 15 sec for timeout settings. Could you provide a list of installed integration in your HA env?

**Comment by djcrawleravp** [comment]
It's ok besides the initial boot this error won't appear anymore

sure, here's a screenshot

<img width="1430" alt="Screen Shot 2023-01-26 at 10 13 44" src="https://user-images.githubusercontent.com/12896341/214857908-2bbb108b-835a-486c-9433-d5e219ad8a2e.png">

**Comment by djcrawleravp** [comment]
Sure, 

<img width="1430" alt="Screen Shot 2023-01-26 at 10 13 44" src="https://user-images.githubusercontent.com/12896341/214858127-9aa37c1d-d758-420d-8cf2-ecd65f37354b.png">

**Comment by djcrawleravp** [comment]
Sure,
<img width="1430" alt="Screen Shot 2023-01-26 at 10 13 44" src="https://user-images.githubusercontent.com/12896341/214858281-c076f15b-8352-4953-aabe-8b95214120ad.png">

**Comment by ollo69** [comment]
@djcrawleravp,

generally this issue is related to something that is slowing down HA during startup. As example, I just discovered that my version of MariaDB had some issue with historical data and was causing lot of errors during start-up (and sometimes also this specific error for this integration). After updating, everything is perfect.
I think that there are no way to fix this at integration level, you should better investigate your HA instance during startup.

**Comment by djcrawleravp** [comment]
@ollo69 

Updating my home assistant docker container to 2023.2 seems to have solved the issue, I haven't' seen the error on the logs since then

**Comment by steveteuber** [comment]
Unfortunately, I have the same issue....

**Environment details:**

Environment (HASSIO, Raspbian, etc): Docker
Home Assistant version installed: 2023.3.1
Component version installed: v0.31.4

**Output of HA logs**
```
Logger: homeassistant.config_entries
Source: custom_components/smartthinq_sensors/wideq/device.py:109
Integration: SmartThinQ LGE Sensors (documentation, issues)
First occurred: 14:35:30 (1 occurrences)
Last logged: 14:35:30

Error setting up entry LGE Devices for smartthinq_sensors
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 383, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/config/custom_components/smartthinq_sensors/__init__.py", line 288, in async_setup_entry
    lge_devices, unsupported_devices, discovered_devices = await lge_devices_setup(
  File "/config/custom_components/smartthinq_sensors/__init__.py", line 570, in lge_devices_setup
    if not await dev.init_device():
  File "/config/custom_components/smartthinq_sensors/__init__.py", line 437, in init_device
    await self._create_coordinator()
  File "/config/custom_components/smartthinq_sensors/__init__.py", line 460, in _create_coordinator
    await coordinator.async_refresh()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 215, in async_refresh
    await self._async_refresh(log_failures=True)
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 239, in _async_refresh
    self.data = await self._async_update_data()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 195, in _async_update_data
    return await self.update_method()
  File "/config/custom_components/smartthinq_sensors/__init__.py", line 465, in _async_update
    await self._async_state_update()
  File "/config/custom_components/smartthinq_sensors/__init__.py", line 477, in _async_state_update
    state = await self._device.poll()
  File "/config/custom_components/smartthinq_sensors/wideq/devices/refrigerator.py", line 360, in poll
    res = await self._device_poll(REFR_ROOT_DATA)
  File "/config/custom_components/smartthinq_sensors/wideq/device.py", line 689, in _device_poll
    snapshot = await self._get_device_snapshot(thinq2_query_device)
  File "/config/custom_components/smartthinq_sensors/wideq/device.py", line 634, in _get_device_snapshot
    return await self._mon.refresh(query_device)
  File "/config/custom_components/smartthinq_sensors/wideq/device.py", line 153, in refresh
    if refresh_auth := await self._refresh_auth():
  File "/config/custom_components/smartthinq_sensors/wideq/device.py", line 109, in _refresh_auth
    async with Monitor._client_lock:
  File "/usr/local/lib/python3.10/asyncio/locks.py", line 14, in __aenter__
    await self.acquire()
  File "/usr/local/lib/python3.10/asyncio/locks.py", line 114, in acquire
    await fut
asyncio.exceptions.CancelledError
```

**Comment by ollo69** [comment]
This issue may be related to limited number of open `aiohttp` connection fixed in `2023.3.4` and in any case, as I understand, it is just a error log that not lock the integration functionality. There is non specific action I can do here, I will leave still open just for possible feed-back with new HA release.

**Comment by steveteuber** [comment]
Thanks for the hint, I've upgraded to `2023.3.4`. I will observe and report back here with logs if necessary.
